src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -1,12 +1,69 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Concurrent;
 using System.Diagnostics.Contracts;
 using System;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Runtime
 {
+    class CoalescedTimeoutTokens
+
+    {
+        private readonly uint _coalescingSpanMs;
+        private readonly ConcurrentDictionary<long, Task<CancellationToken>> _timeToToken =
+            new ConcurrentDictionary<long, Task<CancellationToken>>();
+
+        public CoalescedTimeoutTokens(int coalescingSpanMs = 15)
+        {
+            _coalescingSpanMs = (uint)coalescingSpanMs;
+        }
+
+        public CancellationToken FromTimeout(int millisecondsTimeout)
+        {
+            return FromTimeoutAsync(millisecondsTimeout).Result;
+        }
+
+        public Task<CancellationToken> FromTimeoutAsync(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout <= 0)
+            {
+                return Task.FromResult(new CancellationToken(true));
+            }
+
+            uint currentTime = (uint)Environment.TickCount;
+            long targetTime = millisecondsTimeout + currentTime;
+            targetTime = ((targetTime + (_coalescingSpanMs - 1)) / _coalescingSpanMs) * _coalescingSpanMs;
+
+            Task<CancellationToken> tokenTask;
+
+            if (!_timeToToken.TryGetValue(targetTime, out tokenTask))
+            {
+                var tcs = new TaskCompletionSource<CancellationToken>();
+                if (_timeToToken.TryAdd(targetTime, tcs.Task))
+                {
+                    var token = new CancellationTokenSource((int)(targetTime - currentTime)).Token;
+                    token.Register(t => {
+                        Task<CancellationToken> ignored;
+                        _timeToToken.TryRemove((long)t, out ignored);
+                    }, targetTime);
+                    tcs.TrySetResult(token);
+                    tokenTask = tcs.Task;
```

Comment:
We considered the alternative implementation as followed:

Line 1: if (!s_tokenCache.TryGetValue(targetTime, out tokenTask))
Line 2:    var token = new CancellationTokenSource((int)(targetTime - currentTime)).Token;
Line 3:    if (s_tokenCache.TryAdd(targetTime, token)) .....

What we have observed is that the above implementation still causes enough contention to slow down our high throughput test. There are still a lot of CancellationTokenSource being created at line 2 every 15ms and disposed at the same time.

What we are doing here is using the Task as a placeholder. Note that if multiple threads come in and get false at the line "!s_tokenCache.TryGetValue(targetTime, out tokenTask)", they would all create a tcs, try and insert it to the cache and only 1 will succeed and get to call new CancellationTokenSource(...).